web/src/stream_data.ts

```diff
@@ -267,20 +278,33 @@ export function slug_to_name(slug: string): string {
     */
     const m = /^(\d+)(-.*)?$/.exec(slug);
     if (m) {
-        const stream_id = Number.parseInt(m[1]!, 10);
-        const sub = sub_store.get(stream_id);
-        if (sub) {
-            return sub.name;
+        const stream_id_string = m[1]!;
+        const stream_id = Number.parseInt(stream_id_string, 10);
+        let stream = stream_info.get(stream_id);
+        if (stream) {
+            return stream_id_string;
         }
         // if nothing was found above, we try to match on the stream
         // name in the somewhat unlikely event they had a historical
         // link to a stream like 4-horsemen
+        const stream_name = m[2]!.slice(1); // Remove the "-" at the start
+        stream = get_sub_by_name(stream_name);
+        if (stream) {
+            return stream.stream_id.toString();
+        }
+        // TODO(evy): what should we do here?
+        return slug;
     }
 
     /*
     We are dealing with a pre-2018 slug that doesn't have the
     stream id as a prefix.
     */
+    const stream = get_sub_by_name(slug);
+    if (stream) {
+        return stream.stream_id.toString();
+    }
+    // TODO(evy): what should we do here?
     return slug;
 }
```

Comment:
I think this should probably return `undefined`, and callers should treat this as an invalid stream reference, the same as it would if the integer ID returned after this parsing didn't exist.

Reply:
For this function I recommend taking a look at the mobile code. We converted this area (and basically the rest of the app's internal data structures) to numeric stream IDs a couple of years ago, and the conversion corresponding to this function took some thought but I'm pretty happy with the result.

Here's the version in zulip-mobile:
https://github.com/zulip/zulip-mobile/blob/2217c858e207f9f092651dd853051843c3f04422/src/utils/internalLinks.js#L66-L108

Or for completeness here's the latest version, in zulip-flutter — but it's basically the exact same code translated to Dart:
https://github.com/zulip/zulip-flutter/blob/30bdcc0fd5b949270fc76ddf173ccb4952e287a2/lib/model/internal_link.dart#L262-L296

Two key points are: it can return a nice numeric ID instead of a string; and the pre-2018 case needs a bit of new logic at the end of the function.

Reply:
Oh and here's test cases written for exercising the corner cases of this interaction between the pre-2018 form and the modern form:
https://github.com/zulip/zulip-mobile/blob/2217c858e207f9f092651dd853051843c3f04422/src/utils/__tests__/internalLinks-test.js#L333-L354

Reply:
Thanks, this was helpful! I made a few changes to match what mobile is doing.